### PR TITLE
feat: remove loading indicator when typing in select

### PR DESCRIPTION
**Pull request by ktmud**
### SUMMARY

Remove the loading indicator in select option menus when users are typing and (some) options are already loaded. The loading indicator showing up as users type was added in #16531, while tackling the issue of the inaccurate "No data" message when data is still loading. But we shouldn't show "Loading..." when there are already filtered options available. Removing this unnecessary loading indicator makes the control feels faster.

#### Before

The "Loading..." option shows up as soon as users type, even when all options are loaded:

before-select

#### After

No "Loading..." option when users start to type, but there is a loading indicator in the input and all filtered options will show up after actual loading is finished:

acceptable-select

When all options are loaded for a specific search string, there is no loading indicator at all and no excessive requests will be fired.

### TESTING INSTRUCTIONS

1. Run `npm run storybook` in `superset-frontend`
2. Test it in the storybook: http://localhost:6006/?path=/story/select--async-select
     1. Try search for something in and not in the first page. E.g. "ale", "sandro"
     2. "No data" should not show up if there is a match in loaded options or the next page is still loading
     3. "Loading..." should not show up if there is already a matched option

### ADDITIONAL INFORMATION

- [x] Has associated issue: #16531 #16712 
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/components/Select/Select.tsx
Comment (michael-s-molina) [reply superset-frontend/src/components/Select/Select.tsx:71]: @ktmud The objective of the extension is to change the type of `label` from `React.Node` to `string` and to introduce another property called `customLabel` that accepts a `React.Node`. You can search from `customLabel` to see where it's being used.

**Comment by michael-s-molina** [comment]
@ktmud I found some problems while testing.

The loading indicator is different when we open the select vs when we search for something.

https://user-images.githubusercontent.com/70410625/155366146-d545d13a-d321-42d0-b3a7-de5a21ab269b.mov

The loading spinner is showing after searching and clicking outside the Select. When `fetchOnlyOnSearch` is `true` it never stops spinning.

https://user-images.githubusercontent.com/70410625/155367875-470bdf98-b09c-48be-b6dd-5357bf6d7389.mov

**Comment by ktmud** [comment]
@michael-s-molina Thanks for testing! I've fixed both issues.

File: superset-frontend/src/components/Select/Select.tsx
Comment (ktmud) [reply superset-frontend/src/components/Select/Select.tsx:71]: Then it should extend `OptionData`, not `OptionsProp`.

Comment (ktmud) [reply superset-frontend/src/components/Select/Select.tsx:71]: I'm also curious why did we decide to introduce a `customLabel` prop instead of using [optionRenderer](https://github.com/apache/superset/blob/46715b295c3026c3eb1ee439f1d16fc776e61372/superset-frontend/src/components/Select/DeprecatedSelect.tsx#L87) like in the old Select.

Comment (michael-s-molina) [reply superset-frontend/src/components/Select/Select.tsx:71]: I can't remember any particular reason. I guess it was just a style choice when we were designing the API. I don't see any problem in changing that to `optionRenderer` if you want. It's being used only by 10 files so it would be really easy to migrate.

@geido Please correct me if I'm forgetting anything.

Comment (michael-s-molina) [reply superset-frontend/src/components/Select/Select.tsx:71]: @ktmud Just let me know if you're going to make the change or add the `customLabel` back to the type so I can continue the review.

Comment (ktmud) [reply superset-frontend/src/components/Select/Select.tsx:71]: @michael-s-molina I don't think we should add it back since it's incorrect code.

Comment (michael-s-molina) [reply superset-frontend/src/components/Select/Select.tsx:71]: What I mean is to restore the original intention of the label and customLabel under the correct extended type `OptionData` or introduce the concept of `optionRenderer`. If we merge as it is we don't have any type checking regarding custom labels.

Comment (ktmud) [reply superset-frontend/src/components/Select/Select.tsx:71]: The existing check doesn't work already. I was thinking of some bigger refactoring with the customLabel check so I'd rather keep it in a separate PR. I can put that PR up first if you prefer.

Comment (michael-s-molina) [reply superset-frontend/src/components/Select/Select.tsx:71]: No problem making the change in a follow-up PR. I'll just do a final round of testing tomorrow morning but if I don't find anything we can merge this PR 👍

**Comment by michael-s-molina** [comment]
> @michael-s-molina Thanks for testing! I've fixed both issues.

@ktmud I just tested it again and the bugs are still happening. Did you forget to commit something?

I also noticed that you removed the `debounce` from `handleOnSearch` which is triggering this expensive calculation for every typed key.

https://user-images.githubusercontent.com/70410625/155712543-1342b5fd-d681-41c3-99ff-d4316baac464.mov

**Comment by ktmud** [comment]
@michael-s-molina are you sure you pulled the latest changes? The bugs are fixed for me.

`debounce` is removed for `handleOnSearch` but the actual expensive operations are still debounced as in `debouncedHandleSearch`.

**Comment by michael-s-molina** [comment]
> `debounce` is removed for `handleOnSearch` but the actual expensive operations are still debounced as in `debouncedHandleSearch`.

This part is not debounced:

```
if (allowNewOptions && isSingleMode) {
    const newOption = searchValue &&
      !hasOption(searchValue, selectOptions) && {
        label: searchValue,
        value: searchValue,
        isNewOption: true,
      };
    const cleanSelectOptions = selectOptions.filter(
      opt => !opt.isNewOption,
    );
    const newOptions = newOption
      ? [newOption, ...cleanSelectOptions]
      : cleanSelectOptions;
    setSelectOptions(newOptions);
  }
  if (isAsync && !allValuesLoaded && loadingEnabled) {
    setIsLoading(search !== searchedValue);
  }
```

**Comment by ktmud** [comment]
It's not really expensive.

**Comment by michael-s-molina** [comment]
> @michael-s-molina are you sure you pulled the latest changes? The bugs are fixed for me.

https://user-images.githubusercontent.com/70410625/156048741-53a0281d-740b-400e-a82f-9f10408e7e81.mov

**Comment by michael-s-molina** [comment]
> It's not really expensive. The goal of this whole PR is to make the select options update more responsively to user inputs, which is achieved by separating initial filtering and menu state updates with debounced async requests.

You can see the loading indicator flicking while you type because it's updating the loading state for every key typed.

https://user-images.githubusercontent.com/70410625/156050094-339f24c7-ce3c-4ea7-b9cb-4efdd901cd87.mov

**Comment by ktmud** [comment]
@michael-s-molina I can reproduce the bug now. Looks like this might need a larger refactoring than I thought it needed. Thanks for testing!

**Comment by ktmud** [comment]
@michael-s-molina This should be ready for re-review.

**Review by michael-s-molina — CHANGES_REQUESTED** [review]
The `allowNewOptions` is not working as intended. If you add a new option and then search for it then it will disappear from the options. 

https://user-images.githubusercontent.com/70410625/156366299-28fff4e0-5f79-4629-a1c2-f333f729f76b.mov

File: superset-frontend/src/components/Select/Select.tsx
Comment (ktmud) [reply superset-frontend/src/components/Select/Select.tsx:60]: I was adding `onBlur` then removed it because I changed my implemented. Didn't see `onFocus` used anywhere so removed it, too. Will add it back.

File: superset-frontend/src/components/Select/utils.ts
Comment (michael-s-molina) [inline superset-frontend/src/components/Select/utils.ts:72]:
```suggestion
    optionsArray.find(option =>
      typeof option === 'object' ? option.value === value : option === value,
    ) !== undefined
```

**Review by michael-s-molina — APPROVED** [review]
I haven't found any other problems while testing. Thank you so much @ktmud for improving the user experience of the component and for addressing all suggestions.

@geido was doing some tests with real data so I would kindly request to wait for his approval as well before merging the PR given the critical nature of the component.

**Review by geido — APPROVED** [review]

File: superset-frontend/src/components/Select/utils.ts
Comment (ktmud) [reply superset-frontend/src/components/Select/utils.ts:72]: In general I'd prefer to use `x` in those one liners because sometimes you may want to rename the iterable variable as well.

**Pull request merged**
